### PR TITLE
avoid making substs of type aliases late bound when used as fn args

### LINKED ISSUE #47511: Anonymous bound region BrAnon(0) in return but not args
Looks just like #43567. Mentioning @petrochenkov and @eddyb who were involved in that issue. The following code compiles with 1.21.0 but ICE as of 1.22.0.

```rust
fn f(_: X) -> X {
    unimplemented!()
}

type X<'a> = <&'a () as Trait>::Value;

trait Trait {
    type Value;
}

impl<'a> Trait for &'a () {
    type Value = ();
}

fn main() {}
```

```
error: internal compiler error: librustc_typeck/astconv.rs:1212: anonymous bound region BrAnon(0) in return but not args
 --> src/main.rs:1:15
  |
1 | fn f(_: X) -> X {
  |               ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.25.0-nightly (da569fa9d 2018-01-16) running on x86_64-unknown-linux-gnu
```

### LINKED ISSUE #85533: Intervening type alias of GAT to causes incorrect E0581
This code [[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7299ed4b17607b0f5035ffbb871c96fd):

```rust
#![feature(generic_associated_types)]

trait Trait {
    type Assoc<'a>;
}

type AssocAlias<'a, T> = <T as Trait>::Assoc<'a>;

fn works<'a, T: Trait>(entity: T::Assoc<'a>) -> &'a i32 {&5}
fn fails<'a, T: Trait>(entity: AssocAlias<'a, T>) -> &'a i32 {&5}
```

Produces the following error on 1.54.0-nightly (2021-05-19 f94942d8421dc4b1da86):
```
error[E0581]: return type references lifetime `'a`, which is not constrained by the fn input types
  --> src/lib.rs:10:54
   |
10 | fn fails<'a, T: Trait>(entity: AssocAlias<'a, T>) -> &'a i32 {&5}
   |            
```

The version that uses `T::Assoc` directly works, but the one using the type alias (which should be identical) fails.

### PR DESCRIPTION
**Pull request by BoxyUwU**
fixes #47511 
fixes #85533
(although I did not know theses issues existed when i was working on this :upside_down_face:)

currently `Alias<...>` is treated the same as `Struct<...>` when deciding if generics should be late bound or early bound but this is not correct as `Alias` might normalize to a projection which does not constrain the generics.

I think this needs more tests before merging
more explanation of PR [here](https://hackmd.io/v44a-QVjTIqqhK9uretyQg?view)

Hackmd inline for future readers:
---

This assumes reader is familiar with the concept of early/late bound lifetimes. There's a section on rustc-dev-guide if not (although i think some details are a bit out of date)

## problem & background

Not all lifetimes on a fn can be late bound:
```rust
fn foo<'a>() -> &'a ();
impl<'a> Fn<()> for FooFnDef {
    type Output = &'a (); // uh oh unconstrained lifetime
}
```
so we make make them early bound
```rust
fn foo<'a>() -> &'a ();
impl<'a> Fn<()> for FooFnDef<'a> {// wow look at all that lifetimey
     type Output = &'a ();   
}   
```
(Closures have the same constraint however it is not enforced leading to soundness bugs, [#84385](https://github.com/rust-lang/rust/pull/84385) implements this "downgrading late bound to early bound" for closures)

lifetimes on fn items are only late bound when they are "constrained" by the fn args:
```rust
fn foo<'a>(_: &'a ()) -> &'a ();
//               late bound, not present on `FooFnItem`
//               vv
impl<'a> Trait<(&'a (),)> for FooFnItem {
    type Output = &'a ();    
}

// projections do not constrain inputs
fn bar<'a, T: Trait>(_: <T as Trait<'a>>::Assoc) -> &'a (); //  early bound
                                                            //  vv
impl<'a, T: Trait> Fn<(<T as Trait<'a>>::Assoc,)> for BarFnItem<'a, T> {
    type Output = &'a ();
}
```

current logic for determining if inputs "constrain" a lifetime works off of HIR so does not normalize aliases. It also assumes that any path with no self type constrains all its substs (i.e. `Foo<'a, u32>` has no self type but `T::Assoc` does). This falls apart for top level type aliases (see linked issues):

```rust
type Alias<'a, T> = <T as Trait<'a>>::Assoc;
//                      wow look its a path with no self type uwu
//                      i bet that constrains `'a` so it should be latebound
//                      vvvvvvvvvvv
fn foo<'a, T: Trait>(_: Alias<'a, T>) -> &'a ();
//                     `Alias` normalized to make things clearer
//                     vvvvvvvvvvvvvvvvvvvvvvv
impl<'a, T: Trait> Fn<(<T as Trait<'a>>::Assoc,)> for FooFnDef<T> {
    type Output = &'a (); 
    // oh no `'a` isnt constrained wah wah waaaah *trumbone noises* 
    // i think, idk what musical instrument that is
}
```

## solution

The PR solves this by having the hir visitor that checks for lifetimes in constraining uses check if the path is a `DefKind::Alias`. If it is we ""normalize"" it by calling `type_of` and walking the returned type. This is a bit hacky as it requires a mapping between the substs on the path in hir, and the generics of the `type Alias<...>` which is on the ty layer.

Alternative solutions may involve calculating the "late boundness" of lifetimes after/during astconv rather than relying on hir at all. We already have code to determine whether a lifetime SHOULD be late bound or not as this is currently how the error for `fn foo<'a, T: Trait>(_: Alias<'a, T>) -> &'a ();` gets emitted.

It is probably not possible to do this right now, late boundness is used by `generics_of` and `gather_explicit_predicates_of` as we currently do not put late bound lifetimes in `Generics`. Although this seems sus to me as the long term goal is to make all generics late bound which would result in `generics_of(function)` being empty? [#103448](https://github.com/rust-lang/rust/pull/103448) places all lifetimes in `Generics` regardless of late boundness so that may be a good step towards making this possible.

### CODE REVIEW

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 1]
```diff
@@ -2105,6 +2137,43 @@ fn is_late_bound_map(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<&FxIndexSet<
...
+                    };
+                    walker.visit_ty(self.tcx.type_of(alias_def));
+
+                    match segments.last() {
+                        Some(hir::PathSegment { args: Some(args), .. }) => {
+                            for (_, constrained_arg) in args
+                                .args
+                                .iter()
+                                .enumerate()
+                                .filter(|(n, _)| walker.arg_is_constrained[*n])
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/lifetimes.rs:2167]: You are using the index inside `args` to index a vector which is indexed by a `ty::Generics` index. Those may not match: matching args with params is a 130 line-long code snippet in `astconv::generics`.
However, I'm not sure we could even encounter a mismatch in well-formed code in the arguments to a `TyAlias` (we certainly can on a `Fn`).

File: compiler/rustc_resolve/src/late/lifetimes.rs [hunk 2]
```diff
@@ -2089,12 +2089,44 @@ fn is_late_bound_map(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<&FxIndexSet<
...
+        }
+    }
+
+    struct ConstrainedCollector<'tcx> {
+        tcx: TyCtxt<'tcx>,
         regions: FxHashSet<LocalDefId>,
     }
 
-    impl<'v> Visitor<'v> for ConstrainedCollector {
+    impl<'v> Visitor<'v> for ConstrainedCollector<'v> {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/lifetimes.rs:2129]:
```suggestion
    impl<'v> Visitor<'v> for ConstrainedCollector<'_> {
```
I don't think we need to link the two lifetimes, do we?

File: compiler/rustc_resolve/src/late/lifetimes.rs (hunk 1 above)
Comment (BoxyUwU) [reply compiler/rustc_resolve/src/late/lifetimes.rs:2167]: ah yeah it probably ought to be `.get(*n).unwrap_or_else(|| bug!())` :thinking: I did think that it _probably_ wouldn't be as simple as this logic but wasn't sure what it _should_ be. going to take a look at `astconv::generics`

**Comment by oli-obk** [comment]
This probably needs lang team sign off as it adds potentially a lot more code that compiles.

**Comment by BoxyUwU** [comment]
I'm not sure that this is really "a lot more code that compiles". It would probably be worth getting t-types input to make sure this is not going to mess up any future plans.

File: compiler/rustc_hir_analysis/src/collect/lifetimes.rs [hunk 3]
```diff
@@ -1834,12 +1834,44 @@ fn is_late_bound_map(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<&FxIndexSet<
     debug!(?late_bound);
     return Some(tcx.arena.alloc(late_bound));
 
-    #[derive(Default)]
-    struct ConstrainedCollector {
+    struct ConstrainedCollectorPostAstConv {
```
Comment (nikomatsakis) [inline compiler/rustc_hir_analysis/src/collect/lifetimes.rs:1837]: This needs a lengthy comment linking to the issue and explaining why it exists

Comment (nikomatsakis) [reply compiler/rustc_hir_analysis/src/collect/lifetimes.rs:1837]: And explain why we have to use type-of:

* cross-crate (first)
* would need same basic logic for managing cycles (second)

**Review by nikomatsakis — APPROVED** [review]
r=me modulo the need for an explanatory comment and a cross-crate test

**Comment by nikomatsakis** [comment]
Discussed with @BoxyUwU -- in the fullness of time, I hope this goes away by removing the early/late distinction (and moving everything into generics), but for the time being, I agree with the reasoning of the hackmd that this is the best path forward.

The only real alternative I can see is being very conservative and making everything early bound that appears in a type alias. It seems very likely this will break people and I would be hard pressed to explain to users why it does, so I'm not inclined to go that path unless we find a real complication with this approach beyond that it's a messy hack.

The major concerns I can see are

* it might lead to cycles if it were possible for `type_of` on the type alias to invoke the function, but I don't see why that would happen, at least today
* duplicate code paths that could get out of sync, but this code changes very infrequently

**Pull request merged**